javascript/ql/src/semmle/javascript/dataflow/internal/VariableTypeInference.qll

```diff
@@ -41,6 +41,27 @@ private class SsaVarAccessAnalysis extends DataFlow::AnalyzedValueNode {
   override AbstractValue getALocalValue() { result = def.getAnRhsValue() }
 }
 
+/**
+ * Flow analysis for accesses to SSA variables.
+ *
+ * Unlike `SsaVarAccessAnalysis`, this only contributes to `getAValue()`, not `getALocalValue()`.
+ */
+private class SsaVarAccessWithNonLocalAnalysis extends SsaVarAccessAnalysis {
+  DataFlow::AnalyzedValueNode src;
+
+  SsaVarAccessWithNonLocalAnalysis() {
+    exists(VarDef varDef |
+      varDef = def.(SsaExplicitDefinition).getDef() and
+      varDef.getSource().flow() = src and
+      src instanceof CallWithNonLocalAnalyzedReturnFlow and
+      // avoid relating `v` and `f()` in `var {v} = f();`
+      not varDef.getTarget() instanceof DestructuringPattern
```

Comment:
While I won't let this stand in the way of getting this PR merged, I'd like to express my strong opposition to special-casing this in yet another place. Let's just fix this properly, please?